# SQL Group Prject: ETL (Python)
Kei, Sebastian, Ashlesha, Zach

## <u>STEP 1: Dataset inspection<u>
We first explored and evaluated our non-normalized dataset.

In [1]:
# Import necessary packages
import pandas as pd
from sqlalchemy import create_engine
import psycopg2
import numpy as np
import matplotlib.pyplot as plt
import os

In [2]:
# Load the csv file in a dataframe df:
df = pd.read_csv('Group4_PreNormalization_Dataset.csv', na_values=['#VALUE!', '#DIV/0!', '(blank)'])

/Users/sebastianhubard/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3165: DtypeWarning: Columns (19,21,22,25,26,88) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
df.head()

institution id                                        entity name  year  \
0    800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   
1    800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   
2    800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   
3    800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   
4    800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   

   subgroup cd subgroup name    cohort  cohort count  graduated  grad rate  \
0            1   Multiracial    5-Year             1        0.0          0   
1            1   Multiracial    4-Year             0        0.0          0   
2            2         White  Combined             0        0.0          0   
3            2         White    6-Year             0        0.0          0   
4            2         White    5-Year             2        0.0          0   

   district cd  ... school city  school phone  school grade range  \
0   32070001.0  ...       BRONX    7185858202                6-12   
1   32070001.0  ...       BRONX    7185858202                6-12   
2   32070001.0  ...       BRONX    7185858202                6-12   
3   32070001.0  ...       BRONX    7185858202                6-12   
4   32070001.0  ...       BRONX    7185858202                6-12   

   school num princ multi-princ school  school num teach  \
0                 1                  N                56   
1                 1                  N                56   
2                 1                  N                56   
3                 1                  N                56   
4                 1                  N                56   

  school inexperienced teach  inexperienced teach classification  \
0                       96.0                       Inexperienced   
1                       96.0                       Inexperienced   
2                       96.0                       Inexperienced   
3                       96.0                       Inexperienced   
4                       96.0                       Inexperienced   

  school num counselors school num social  
0                     2                 0  
1                     2                 0  
2                     2                 0  
3                     2                 0  
4                     2                 0  

[5 rows x 95 columns]

### Rename column names

In [4]:
# print df column names (original column names in pre-3NF dataset)
print(df.columns)

Index(['institution id', 'entity name', 'year', 'subgroup cd', 'subgroup name',
       'cohort', 'cohort count', 'graduated', 'grad rate', 'district cd',
       'district name', 'boces cd', 'boces name', 'county cd', 'county name',
       'needs index', 'needs desc', 'overall status cd', 'overall status',
       'change status flag', 'pupil count tot', 'receives federal exp',
       'federal funding level', 'federal funding', 'per federal exp',
       'receives state local exp', 'state local funding level',
       'state local exp', 'per state local exp', 'fed state local exp',
       'per fed state local exp', 'group code', 'group name', 'pre k',
       'kindergarten', 'grade 1', 'grade 2', 'grade 3', 'grade 4', 'grade 5',
       'grade 6', 'grade 7', 'grade 8', 'grade 9', 'grade 10', 'grade 11',
       'grade 12', 'k12', 'num ell', 'per ell', 'num am ind', 'per am ind',
       'num black', 'per black', 'num hisp', 'per hisp', 'num asian',
       'per asian', 'num white', 'per white',

In [5]:
# mass edit column names of df to match the attribute names in our db plan for easier mapping during ETL
df = df.set_axis(['institution_id', 'entity_name', 'year', 'subgroup_cd', 'subgroup_name',
       'cohort_descr', 'cohort_count', 'graduated', 'grad_rate', 'school_district_id',
       'school_district_name', 'boces_cd', 'boces_name', 'county_cd', 'county_name',
       'school_fundingtype_id', 'school_fundingtype_descr', 'overall_status_cd', 'overall_status',
       'change_status_flag', 'pupil_count_tot', 'receives_federal_exp',
       'federal_funding_level', 'federal_funding', 'per_federal_exp',
       'receives_state_local_exp', 'state_local_funding_level',
       'state_local_exp', 'per_state_local_exp', 'fed_state_local_exp',
       'per_fed_state_local_exp', 'group_code', 'group_name', 'pre_k',
       'kindergarten', 'grade_1', 'grade_2', 'grade_3', 'grade_4', 'grade_5',
       'grade_6', 'grade_7', 'grade_8', 'grade_9', 'grade_10', 'grade_11',
       'grade_12', 'k12', 'num_ell', 'per_ell', 'num_am_ind', 'per_am_ind',
       'num_black', 'per_black', 'num_hisp', 'per_hisp', 'num_asian',
       'per_asian', 'num_white', 'per_white', 'num_multi', 'per_multi',
       'num_swd', 'per_swd', 'num_female', 'per_female', 'num_male',
       'per_male', 'num_ecdis', 'per_ecdis', 'num_homeless', 'per_homeless',
       'num_foster', 'per_foster', 'num_armed', 'per_armed', 'num_suspensions',
       'per_suspensions', 'attendance_rate', 'attendance_classification',
       'num_free_lunch', 'per_free_lunch', 'num_reduced_lunch',
       'per_reduced_lunch', 'school_street', 'school_city', 'school_phone',
       'school_grade_range', 'school_num_princ', 'multi_princ_school',
       'school_num_teach', 'school_inexperienced_teach',
       'inexperienced_teach_classification', 'school_num_counselors',
       'school_num_social'], axis=1)

In [6]:
# print updated df column names
print(df.columns)

Index(['institution_id', 'entity_name', 'year', 'subgroup_cd', 'subgroup_name',
       'cohort_descr', 'cohort_count', 'graduated', 'grad_rate',
       'school_district_id', 'school_district_name', 'boces_cd', 'boces_name',
       'county_cd', 'county_name', 'school_fundingtype_id',
       'school_fundingtype_descr', 'overall_status_cd', 'overall_status',
       'change_status_flag', 'pupil_count_tot', 'receives_federal_exp',
       'federal_funding_level', 'federal_funding', 'per_federal_exp',
       'receives_state_local_exp', 'state_local_funding_level',
       'state_local_exp', 'per_state_local_exp', 'fed_state_local_exp',
       'per_fed_state_local_exp', 'group_code', 'group_name', 'pre_k',
       'kindergarten', 'grade_1', 'grade_2', 'grade_3', 'grade_4', 'grade_5',
       'grade_6', 'grade_7', 'grade_8', 'grade_9', 'grade_10', 'grade_11',
       'grade_12', 'k12', 'num_ell', 'per_ell', 'num_am_ind', 'per_am_ind',
       'num_black', 'per_black', 'num_hisp', 'per_hisp', 'num_as

### Check for null values

In [7]:
from pprint import pprint
print('Count of Attributes with NULL values:',len(df.columns[df.isna().any()].tolist()))

# For loop to print out each column of the `naAttributes` list
naAttributes = df.columns[df.isna().any()].tolist()
for col in range(len(naAttributes)):
    print(naAttributes[col])

Count of Attributes with NULL values: 17
school_district_id
school_district_name
boces_name
change_status_flag
pupil_count_tot
receives_federal_exp
federal_funding_level
federal_funding
per_federal_exp
receives_state_local_exp
state_local_funding_level
state_local_exp
per_state_local_exp
fed_state_local_exp
per_fed_state_local_exp
school_inexperienced_teach
inexperienced_teach_classification


**Based on the above, there are 17 attributes with NULL values. In anticipation of loading our dataset to the db, we will not enforce NULL constraints for non-primary key attributes to prevent data loss during ETL.**

### Check data types of attributes

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22080 entries, 0 to 22079
Data columns (total 95 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   institution_id                      22080 non-null  int64  
 1   entity_name                         22080 non-null  object 
 2   year                                22080 non-null  int64  
 3   subgroup_cd                         22080 non-null  int64  
 4   subgroup_name                       22080 non-null  object 
 5   cohort_descr                        22080 non-null  object 
 6   cohort_count                        22080 non-null  int64  
 7   graduated                           22080 non-null  float64
 8   grad_rate                           22080 non-null  int64  
 9   school_district_id                  19520 non-null  float64
 10  school_district_name                19520 non-null  object 
 11  boces_cd                            22080

**From the code above, we see that there are 22,079 rows in the dataframe and 95 columns. We also see we need to change the data type of the following attributes:**
* institution_id	char(12)
* year	char(4)
* subgroup_cd	varchar(2)
* school_district_id	char(8)
* county_cd	char(2)
* overall_status_cd	varchar(2)
* group_code	varchar(2)
* school_phone	char(10)

In [9]:
# Attributes to change to object (string)
df['institution_id']=df['institution_id'].astype('object')
df['year']=df['year'].astype('object')
df['subgroup_cd']=df['subgroup_cd'].astype('object')
df['school_district_id']=df['school_district_id'].astype('object')
df['county_cd']=df['county_cd'].astype('object')
df['overall_status_cd']=df['overall_status_cd'].astype('object')
df['group_code']=df['group_code'].astype('object')
df['school_phone']=df['school_phone'].astype('object')

# check data types using df.dtypes method 
print(df.dtypes)

institution_id                         object
entity_name                            object
year                                   object
subgroup_cd                            object
subgroup_name                          object
                                       ...   
school_num_teach                        int64
school_inexperienced_teach            float64
inexperienced_teach_classification     object
school_num_counselors                   int64
school_num_social                       int64
Length: 95, dtype: object


## <u>STEP 2: Create database tables in PostgreSQL db</u>  
**NOTE:** Prior to creating the database tables, we finalized our normalization plan and ER Diagram (Checkpoint 3). In this step, we will create the database tables using Python & SQL.

### 2a) Create `schools` database in pgAdmin

Prior to running the code below, we first created an `education` database in pgAdmin

### 2b) Create db tables

#### Connect to PGAdmin

In [10]:
# Make sure to create `schools` database directly in pgAdmin prior to running the following code

engine = create_engine("postgresql://postgres:123@localhost:5432/schools")
connection = engine.connect()

In [11]:
# SQL Code to create the 3NF empty tables (in alignment with ER diagram)
stmt = """

    CREATE TABLE county(
        county_cd char(2) PRIMARY KEY NOT NULL,
        county_name VARCHAR(30) NOT NULL
    );

    CREATE TABLE school_district (
        school_district_id varchar(100) PRIMARY KEY NOT NULL,
        school_district_name varchar(100)
    );

    CREATE TABLE cohort (
        cohort_id numeric PRIMARY KEY NOT NULL,
        cohort_descr varchar(10)
    );

    CREATE TABLE subgroup(
        subgroup_cd VARCHAR(2) PRIMARY KEY NOT NULL,
        subgroup_name VARCHAR(50) NOT NULL
    );

    CREATE TABLE county_schooldistrict (
        school_district_id varchar(100) NOT NULL,
        county_cd char(2) NOT NULL,
        PRIMARY KEY (county_cd, school_district_id),
		FOREIGN KEY (county_cd) REFERENCES county(county_cd),
		FOREIGN KEY (school_district_id) REFERENCES school_district(school_district_id)
    );

    CREATE TABLE institution(
        institution_id char(12) PRIMARY KEY NOT NULL,
        school_district_id varchar(100), 
        FOREIGN KEY (school_district_id) REFERENCES school_district(school_district_id),
        boces_cd char(4), 
        county_cd char(2), 
        FOREIGN KEY (county_cd) REFERENCES county(county_cd),
        school_fundingtype_descr VARCHAR(30), 
        group_name VARCHAR(20),
        entity_name VARCHAR(100),
        school_street VARCHAR(60),
        school_city VARCHAR(60),
        school_phone char(10),
        school_grade_range VARCHAR(60),
		FOREIGN KEY(county_cd) REFERENCES county(county_cd)     
    );

    CREATE TABLE institution_annualrates(
        institution_id char(12) NOT NULL,
            FOREIGN KEY (institution_id) REFERENCES institution(institution_id),
        year char(4) NOT NULL,
        pupil_count_tot NUMERIC,
        federal_funding_level VARCHAR(30), 
        federal_funding NUMERIC,
        per_federal_exp NUMERIC,
        state_local_funding_level VARCHAR(30), 
        state_local_exp NUMERIC,
        per_state_local_exp NUMERIC,
        fed_state_local_exp NUMERIC,
        per_fed_state_local_exp NUMERIC,
        attendance_rate NUMERIC, 
		attendance_classification VARCHAR(20),
        pre_k NUMERIC,
        kindergarten NUMERIC,
        grade_1 NUMERIC,
        grade_2 NUMERIC,
        grade_3 NUMERIC,
        grade_4 NUMERIC,
        grade_5 NUMERIC,
        grade_6 NUMERIC,
        grade_7 NUMERIC,
        grade_8 NUMERIC,
        grade_9 NUMERIC,
        grade_10 NUMERIC,
        grade_11 NUMERIC,
        grade_12 NUMERIC,
        k12 NUMERIC,
        num_ell NUMERIC,
        per_ell NUMERIC,
        num_am_ind NUMERIC,
        per_am_ind NUMERIC,
        num_black NUMERIC,
        per_black NUMERIC,
        num_hisp NUMERIC,
        per_hisp NUMERIC,
        num_asian NUMERIC,
        per_asian NUMERIC,
        num_white NUMERIC,
        per_white NUMERIC,
        num_multi NUMERIC,
        per_multi NUMERIC,
        num_swd NUMERIC,
        per_swd NUMERIC,
        num_female NUMERIC,
        per_female NUMERIC,
        num_male NUMERIC,
        per_male NUMERIC,
        num_ecdis NUMERIC,
        per_ecdis NUMERIC,
        num_homeless NUMERIC,
        per_homeless NUMERIC,
        num_foster NUMERIC,
        per_foster NUMERIC,
        num_armed NUMERIC,
        per_armed NUMERIC,
        num_suspensions NUMERIC,
        per_suspensions NUMERIC,
        num_free_lunch NUMERIC,
        per_free_lunch NUMERIC,
        num_reduced_lunch NUMERIC,
        per_reduced_lunch NUMERIC,
        school_inexperienced_teach NUMERIC, 
        inexperienced_teach_classification VARCHAR(30),
        school_num_princ VARCHAR(2), 
        school_num_counselors NUMERIC,
        school_num_social NUMERIC,
        overall_status VARCHAR(100),
        CONSTRAINT PK_inst_yr 
            PRIMARY KEY (institution_id, year)
    );

    CREATE TABLE subgroup_cohorts_annualrates(
        institution_id char(12) NOT NULL,
        year char(4) NOT NULL,
        subgroup_cd VARCHAR(2) NOT NULL,
        cohort_id numeric NOT NULL,
        cohort_count NUMERIC, 
        graduated NUMERIC,
        grad_rate NUMERIC,
        CONSTRAINT PK_inst_yr_coh_subgr 
            PRIMARY KEY (institution_id, year, subgroup_cd, cohort_id),
        CONSTRAINT fk_institution
            FOREIGN KEY (institution_id) REFERENCES institution(institution_id),
      --  CONSTRAINT fk_institution_annualrates
      --      FOREIGN KEY (institution_id, year) REFERENCES institution_annualrates(institution_id, year),
        CONSTRAINT fk_subgroup_cohort_subgroup
            FOREIGN KEY (subgroup_cd) REFERENCES subgroup(subgroup_cd),
        CONSTRAINT fk_subgroup_cohort_cohort
            FOREIGN KEY (cohort_id) REFERENCES cohort(cohort_id)
    );   
	
	CREATE TABLE teachers(
		-- a teacher may not work for more than one institution in the same year
        institution_id char(12) NOT NULL,
        year char(4) NOT NULL,
        teacher_id VARCHAR(6) NOT NULL,
        teacher_name VARCHAR(60) NOT NULL,
        PRIMARY KEY (teacher_id, year),
        CONSTRAINT fk_teachers
            FOREIGN KEY (institution_id) REFERENCES institution(institution_id)
	); 
	
	CREATE TABLE students(
		-- a student may not attend for more than one institution in the same year
        institution_id char(12) NOT NULL,
        year char(4) NOT NULL,
        student_id VARCHAR(6) NOT NULL,
        student_name VARCHAR(60) NOT NULL,
        PRIMARY KEY (student_id, year),
        CONSTRAINT fk_students
            FOREIGN KEY (institution_id) REFERENCES institution(institution_id)
	);
	
	CREATE TABLE student_advisors(
		-- a student may not attend for more than one institution in the same year
        year char(4) NOT NULL,
        student_id VARCHAR(6) NOT NULL,
        advisor_id VARCHAR(6) NOT NULL,
        PRIMARY KEY (student_id, year),
        CONSTRAINT fk_advisors_students
            FOREIGN KEY (student_id, year) REFERENCES students(student_id, year),
		CONSTRAINT fk_advisors_teachers
            FOREIGN KEY (advisor_id, year) REFERENCES teachers(teacher_id, year)
	);
	
	CREATE TABLE taught_by_history(
		-- a student may not attend for more than one institution in the same year
        institution_id char(12) NOT NULL,
        year char(4) NOT NULL,        
		student_id VARCHAR(6) NOT NULL,
        teacher_id VARCHAR(6) NOT NULL,
		course_id VARCHAR(6) NOT NULL,
        PRIMARY KEY (student_id, year, course_id),
        CONSTRAINT fk_taught_students
            FOREIGN KEY (student_id, year) REFERENCES students(student_id, year),
		CONSTRAINT fk_taught_teachers
            FOREIGN KEY (teacher_id, year) REFERENCES teachers(teacher_id, year)
	);
	
	CREATE TABLE classrooms(
        institution_id char(12) NOT NULL,
        building VARCHAR(60) NOT NULL,        
		classroom VARCHAR(60) NOT NULL,
        PRIMARY KEY (institution_id, building, classroom),
        CONSTRAINT fk_classroom_institution
            FOREIGN KEY (institution_id) REFERENCES institution(institution_id)
	);
	
	CREATE TABLE homerooms(
		-- each teacher can only have one homeroom a year
        institution_id char(12) NOT NULL,
		year char(4) NOT NULL,        
        building VARCHAR(60) NOT NULL,        
		classroom VARCHAR(60) NOT NULL,
		teacher_id VARCHAR(6) NOT NULL,
        PRIMARY KEY (year, teacher_id),
        CONSTRAINT fk_homeroom_institution
            FOREIGN KEY (institution_id) REFERENCES institution(institution_id),
		CONSTRAINT fk_homeroom_teacher
            FOREIGN KEY (teacher_id, year) REFERENCES teachers(teacher_id, year)
	);

"""

# Execute the statement to create empty 3NF tables
connection.execute(stmt)

***

## <u>STEP 3: Extract, Transform, Load (ETL)</u>  

In [12]:
# making a df backup in case we need to revert to earlier changes
df_backup_1 = df
df_backup_1.head(1)

institution_id                                        entity_name  year  \
0   800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   

  subgroup_cd subgroup_name cohort_descr  cohort_count  graduated  grad_rate  \
0           1   Multiracial       5-Year             1        0.0          0   

  school_district_id  ... school_city  school_phone  school_grade_range  \
0         32070001.0  ...       BRONX    7185858202                6-12   

  school_num_princ multi_princ_school  school_num_teach  \
0                1                  N                56   

  school_inexperienced_teach inexperienced_teach_classification  \
0                       96.0                      Inexperienced   

  school_num_counselors school_num_social  
0                     2                 0  

[1 rows x 95 columns]

### 3a) Populate table `subgroup_cohorts_annualrates`

In [13]:
sub_coh_annual_temp_df = df[['institution_id', 'year', 'subgroup_cd', 'cohort_descr', 'cohort_count', 'graduated', 'grad_rate']]. \
    drop_duplicates()

# drop rows with null values where indicated NOT NULL in the table parameters
sub_coh_annual_temp_df = sub_coh_annual_temp_df.dropna(axis=0, how="any", thresh=None, inplace=False)

sub_coh_annual_temp_df.head()

institution_id  year subgroup_cd cohort_descr  cohort_count  graduated  \
0   800000057113  2020           1       5-Year             1        0.0   
1   800000057113  2020           1       4-Year             0        0.0   
2   800000057113  2020           2     Combined             0        0.0   
3   800000057113  2020           2       6-Year             0        0.0   
4   800000057113  2020           2       5-Year             2        0.0   

   grad_rate  
0          0  
1          0  
2          0  
3          0  
4          0

### 3b) Populate table `cohort`

In [14]:
# Populate table with just unique cohort_descr values
cohort_temp_df = df[['cohort_descr']]. \
    drop_duplicates()

# Add a cohort_id
cohort_temp_df.insert(0, 'cohort_id', range(1, 1 + len(cohort_temp_df)))

# drop rows with null values where indicated NOT NULL in the table parameters
cohort_temp_df = cohort_temp_df.dropna(axis=0, how="any", subset = ['cohort_id'], thresh=None, inplace=False)

cohort_temp_df.head()

cohort_id cohort_descr
0          1       5-Year
1          2       4-Year
2          3     Combined
3          4       6-Year

In [15]:
# Map cohort ID back to the subgroup_cohorts_annualrates table
cohort_id_list = [cohort_temp_df.cohort_id[cohort_temp_df.cohort_descr == i].values[0] for i in df.cohort_descr]
sub_coh_annual_temp_df.insert(4, 'cohort_id', cohort_id_list)

# Remove cohort_descr from the subgroup_cohorts_annualrates table
del sub_coh_annual_temp_df['cohort_descr']

sub_coh_annual_temp_df.head()

institution_id  year subgroup_cd  cohort_id  cohort_count  graduated  \
0   800000057113  2020           1          1             1        0.0   
1   800000057113  2020           1          2             0        0.0   
2   800000057113  2020           2          3             0        0.0   
3   800000057113  2020           2          4             0        0.0   
4   800000057113  2020           2          1             2        0.0   

   grad_rate  
0          0  
1          0  
2          0  
3          0  
4          0

### 3c) Populate table `subgroup`

In [16]:
# create a temporary table with a unique combination of subgroup_cd and subgroup_name
subgroup_temp_df = df.groupby(['subgroup_cd','subgroup_name']).size().reset_index().rename(columns={0:'count'})
del subgroup_temp_df['count']

# drop rows with null values where indicated NOT NULL in the table parameters
subgroup_temp_df = subgroup_temp_df.dropna(axis=0, how="any", subset = ['subgroup_cd'], thresh=None, inplace=False)

subgroup_temp_df.head()

subgroup_cd               subgroup_name
0            1                 Multiracial
1            2                       White
2            3          Hispanic or Latino
3            4  Economically Disadvantaged
4            5   English Language Learners

In [17]:
# Remove subgroup and cohort columns from df
del df['subgroup_cd']
del df['subgroup_name']
del df['cohort_descr']
del df['cohort_count']
del df['graduated']
del df['grad_rate']
df.head()

institution_id                                        entity_name  year  \
0   800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   
1   800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   
2   800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   
3   800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   
4   800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   

  school_district_id                   school_district_name  boces_cd  \
0         32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   
1         32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   
2         32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   
3         32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   
4         32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   

   boces_name county_cd county_name  school_fundingtype_id  ... school_city  \
0         NaN        32       BRONX                      1  ...       BRONX   
1         NaN        32       BRONX                      1  ...       BRONX   
2         NaN        32       BRONX                      1  ...       BRONX   
3         NaN        32       BRONX                      1  ...       BRONX   
4         NaN        32       BRONX                      1  ...       BRONX   

  school_phone school_grade_range school_num_princ  multi_princ_school  \
0   7185858202               6-12                1                   N   
1   7185858202               6-12                1                   N   
2   7185858202               6-12                1                   N   
3   7185858202               6-12                1                   N   
4   7185858202               6-12                1                   N   

  school_num_teach school_inexperienced_teach  \
0               56                       96.0   
1               56                       96.0   
2               56                       96.0   
3               56                       96.0   
4               56                       96.0   

   inexperienced_teach_classification  school_num_counselors school_num_social  
0                       Inexperienced                      2                 0  
1                       Inexperienced                      2                 0  
2                       Inexperienced                      2                 0  
3                       Inexperienced                      2                 0  
4                       Inexperienced                      2                 0  

[5 rows x 89 columns]

In [18]:
# remove subgroup duplicative rows from df
df = df.drop_duplicates()

df.head()

institution_id                                        entity_name  year  \
0     800000057113  MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...  2020   
40    800000058074   SOUTH BRONX PREPARATORY - A COLLEGE BOARD SCHOOL  2020   
80    800000077812  HERO (HEALTH, EDUCATION AND RESEARCH OCCUPATIO...  2020   
120   800000059630                INTERNATIONAL COMMUNITY HIGH SCHOOL  2020   
160   800000060401                 JILL CHAIFETZ TRANSFER HIGH SCHOOL  2020   

    school_district_id                   school_district_name  boces_cd  \
0           32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   
40          32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   
80          32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   
120         32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   
160         32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7      3090   

     boces_name county_cd county_name  school_fundingtype_id  ... school_city  \
0           NaN        32       BRONX                      1  ...       BRONX   
40          NaN        32       BRONX                      1  ...       BRONX   
80          NaN        32       BRONX                      1  ...       BRONX   
120         NaN        32       BRONX                      1  ...       BRONX   
160         NaN        32       BRONX                      1  ...       BRONX   

    school_phone school_grade_range school_num_princ  multi_princ_school  \
0     7185858202               6-12                1                   N   
40    7182922211           6-12, US                1                   N   
80    7185858013           9-12, US                1                   N   
120   7186654128           9-12, US                1                   N   
160   7184022429               9-12                1                   N   

    school_num_teach school_inexperienced_teach  \
0                 56                       96.0   
40                59                       83.0   
80                44                      118.0   
120               26                       88.0   
160               20                      150.0   

     inexperienced_teach_classification  school_num_counselors  \
0                         Inexperienced                      2   
40               Moderately Experienced                      2   
80                        Inexperienced                      0   
120              Moderately Experienced                      2   
160                       Inexperienced                      0   

    school_num_social  
0                   0  
40                  0  
80                  1  
120                 1  
160                 1  

[5 rows x 89 columns]

### 3d) Populate table `Institution`

In [19]:
institution_temp_df = df[['institution_id', 'school_district_id', 'boces_cd', 'county_cd', 'school_fundingtype_descr', 'group_name', 'entity_name', 'school_street', 'school_city', 'school_phone', 'school_grade_range']]. \
    drop_duplicates()

# drop rows with null values where indicated NOT NULL in the table parameters
institution_temp_df = institution_temp_df.dropna(axis=0, how="any", subset = ['institution_id'], thresh=None, inplace=False)

institution_temp_df.head()

institution_id school_district_id  boces_cd county_cd  \
0     800000057113         32070001.0      3090        32   
40    800000058074         32070001.0      3090        32   
80    800000077812         32070001.0      3090        32   
120   800000059630         32070001.0      3090        32   
160   800000060401         32070001.0      3090        32   

    school_fundingtype_descr     group_name  \
0                     Public  Public School   
40                    Public  Public School   
80                    Public  Public School   
120                   Public  Public School   
160                   Public  Public School   

                                           entity_name      school_street  \
0    MS 223 LABORATORY SCHOOL OF FINANCE AND TECHNO...     360 E 145TH ST   
40    SOUTH BRONX PREPARATORY - A COLLEGE BOARD SCHOOL     360 E 145TH ST   
80   HERO (HEALTH, EDUCATION AND RESEARCH OCCUPATIO...  455 SOUTHERN BLVD   
120                INTERNATIONAL COMMUNITY HIGH SCHOOL      345 BROOK AVE   
160                 JILL CHAIFETZ TRANSFER HIGH SCHOOL     778 FOREST AVE   

    school_city school_phone school_grade_range  
0         BRONX   7185858202               6-12  
40        BRONX   7182922211           6-12, US  
80        BRONX   7185858013           9-12, US  
120       BRONX   7186654128           9-12, US  
160       BRONX   7184022429               9-12

### 3e) Populate table `institution_annualrates`

In [21]:
numericrate_df = df[['institution_id','year','pupil_count_tot','federal_funding_level','federal_funding','per_federal_exp','state_local_funding_level','state_local_exp','per_state_local_exp','fed_state_local_exp','per_fed_state_local_exp','attendance_rate','attendance_classification','pre_k','kindergarten','grade_1','grade_2','grade_3','grade_4','grade_5','grade_6','grade_7','grade_8','grade_9','grade_10','grade_11','grade_12','k12','num_ell','per_ell','num_am_ind','per_am_ind','num_black','per_black','num_hisp','per_hisp','num_asian','per_asian','num_white','per_white','num_multi','per_multi','num_swd','per_swd','num_female','per_female','num_male','per_male','num_ecdis','per_ecdis','num_homeless','per_homeless','num_foster','per_foster','num_armed','per_armed','num_suspensions','per_suspensions','num_free_lunch','per_free_lunch','num_reduced_lunch','per_reduced_lunch','school_inexperienced_teach','inexperienced_teach_classification','school_num_princ','school_num_counselors','school_num_social','overall_status']]

In [22]:
numericrate_df2 = numericrate_df.drop_duplicates(subset=None, keep='first', inplace=False)

In [23]:
numericrate_df2.head()

institution_id  year  pupil_count_tot federal_funding_level  \
0     800000057113  2020            688.0          $100K to $1M   
40    800000058074  2020            650.0          $100K to $1M   
80    800000077812  2020            536.0          $100K to $1M   
120   800000059630  2020            386.0          $100K to $1M   
160   800000060401  2020            165.0         $10k to $100k   

     federal_funding  per_federal_exp state_local_funding_level  \
0           678377.0            986.0               $1M or more   
40          939744.0           1446.0               $1M or more   
80          372923.0            696.0               $1M or more   
120         552330.0           1431.0               $1M or more   
160          88913.0            539.0               $1M or more   

     state_local_exp  per_state_local_exp  fed_state_local_exp  ...  \
0          8827896.0              12831.0            9506273.0  ...   
40         9299284.0              14307.0           10239028.0  ...   
80         7562112.0              14108.0            7935035.0  ...   
120        6537056.0              16935.0            7089386.0  ...   
160        3753682.0              22750.0            3842595.0  ...   

     num_free_lunch  per_free_lunch num_reduced_lunch  per_reduced_lunch  \
0               628              91                14                  2   
40              612              94                 6                  1   
80              446              83                 5                  1   
120             381              99                 1                  0   
160             145              88                 0                  0   

     school_inexperienced_teach  inexperienced_teach_classification  \
0                          96.0                       Inexperienced   
40                         83.0              Moderately Experienced   
80                        118.0                       Inexperienced   
120                        88.0              Moderately Experienced   
160                       150.0                       Inexperienced   

     school_num_princ  school_num_counselors  school_num_social  \
0                   1                      2                  0   
40                  1                      2                  0   
80                  1                      0                  1   
120                 1                      2                  1   
160                 1                      0                  1   

                            overall_status  
0                            Good Standing  
40                           Good Standing  
80                           Good Standing  
120                          Good Standing  
160  Comprehensive Support and Improvement  

[5 rows x 68 columns]

In [24]:
#REMOVE DUPLICATE - This row was still a duplicate even after the above drop this confirms it
data = numericrate_df2.drop(labels=16360, axis=0)
print(data[data['institution_id'] == 800000046813])

      institution_id  year  pupil_count_tot federal_funding_level  \
16384   800000046813  2020            517.0                   NaN   

       federal_funding  per_federal_exp state_local_funding_level  \
16384         286237.0            554.0                       NaN   

       state_local_exp  per_state_local_exp  fed_state_local_exp  ...  \
16384        6872182.0              13292.0            7158419.0  ...   

       num_free_lunch  per_free_lunch num_reduced_lunch  per_reduced_lunch  \
16384             383              74                30                  6   

       school_inexperienced_teach  inexperienced_teach_classification  \
16384                        65.0                         Experienced   

       school_num_princ  school_num_counselors  school_num_social  \
16384                 1                      2                  1   

       overall_status  
16384   Good Standing  

[1 rows x 68 columns]


In [25]:
institution_annualrates_temp_df = data
institution_annualrates_temp_df.head()

institution_id  year  pupil_count_tot federal_funding_level  \
0     800000057113  2020            688.0          $100K to $1M   
40    800000058074  2020            650.0          $100K to $1M   
80    800000077812  2020            536.0          $100K to $1M   
120   800000059630  2020            386.0          $100K to $1M   
160   800000060401  2020            165.0         $10k to $100k   

     federal_funding  per_federal_exp state_local_funding_level  \
0           678377.0            986.0               $1M or more   
40          939744.0           1446.0               $1M or more   
80          372923.0            696.0               $1M or more   
120         552330.0           1431.0               $1M or more   
160          88913.0            539.0               $1M or more   

     state_local_exp  per_state_local_exp  fed_state_local_exp  ...  \
0          8827896.0              12831.0            9506273.0  ...   
40         9299284.0              14307.0           10239028.0  ...   
80         7562112.0              14108.0            7935035.0  ...   
120        6537056.0              16935.0            7089386.0  ...   
160        3753682.0              22750.0            3842595.0  ...   

     num_free_lunch  per_free_lunch num_reduced_lunch  per_reduced_lunch  \
0               628              91                14                  2   
40              612              94                 6                  1   
80              446              83                 5                  1   
120             381              99                 1                  0   
160             145              88                 0                  0   

     school_inexperienced_teach  inexperienced_teach_classification  \
0                          96.0                       Inexperienced   
40                         83.0              Moderately Experienced   
80                        118.0                       Inexperienced   
120                        88.0              Moderately Experienced   
160                       150.0                       Inexperienced   

     school_num_princ  school_num_counselors  school_num_social  \
0                   1                      2                  0   
40                  1                      2                  0   
80                  1                      0                  1   
120                 1                      2                  1   
160                 1                      0                  1   

                            overall_status  
0                            Good Standing  
40                           Good Standing  
80                           Good Standing  
120                          Good Standing  
160  Comprehensive Support and Improvement  

[5 rows x 68 columns]

### 3f) Populate table `county`

In [26]:
county_temp_df = df[['county_cd', 'county_name']]. \
    drop_duplicates()

# drop rows with null values where indicated NOT NULL in the table parameters
county_temp_df = county_temp_df.dropna(axis=0, how="any", thresh=None, inplace=False)

county_temp_df.head()

county_cd county_name
0            32       BRONX
5640         33       KINGS
12080        31    NEW YORK
17920        34      QUEENS
21560        35    RICHMOND

### 3g) Populate table `school_district`

In [27]:
school_district_temp_df = df[['school_district_id', 'school_district_name']]. \
    drop_duplicates()

# drop rows with null values where indicated NOT NULL in the table parameters
school_district_temp_df = school_district_temp_df.dropna(axis=0, how="any", thresh=None, inplace=False)

# #another to get rid of NaN
# df = df[df.school_district_id != "NaN"]

school_district_temp_df = school_district_temp_df.drop_duplicates(subset='school_district_id', keep="first")

school_district_temp_df.head()

school_district_id                   school_district_name
0            32070001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 7
920          32080001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 8
1720         32090001.0  NEW YORK CITY GEOGRAPHIC DISTRICT # 9
2720         32100001.0  NEW YORK CITY GEOGRAPHIC DISTRICT #10
4080         32110001.0  NEW YORK CITY GEOGRAPHIC DISTRICT #11

### 3h) Populate table `county_schooldistrict`

In [28]:
county_schooldistrict_temp_df = df[['school_district_id', 'county_cd']]. \
    drop_duplicates()

# drop rows with null values where indicated NOT NULL in the table parameters
county_schooldistrict_temp_df = county_schooldistrict_temp_df.dropna(axis=0, how="any", thresh=None, inplace=False)

county_schooldistrict_temp_df.head()

school_district_id county_cd
0            32070001.0        32
920          32080001.0        32
1720         32090001.0        32
2720         32100001.0        32
4080         32110001.0        32

## <u> STEP 4: Load all temporary tables to PGAdmin (PostrgeSQL 'schools' database) <u>

#### Because the tables have to be pushed in a certain order per their primary/foreign key relations, and we have decided to not necessarily create the tables in that order, we have decided to push the tables at the very end in the correct order:

In [29]:
county_temp_df.to_sql(name='county', con=engine, if_exists='append', index=False)

In [30]:
school_district_temp_df.to_sql(name='school_district', con=engine, if_exists='append', index=False)

In [31]:
cohort_temp_df.to_sql(name='cohort', con=engine, if_exists='append', index=False)

In [32]:
subgroup_temp_df.to_sql(name='subgroup', con=engine, if_exists='append', index=False)

In [33]:
county_schooldistrict_temp_df.to_sql(name='county_schooldistrict', con=engine, if_exists='append', index=False)

In [34]:
institution_temp_df.to_sql(name='institution', con=engine, if_exists='append', index=False)

In [35]:
institution_annualrates_temp_df.to_sql(name='institution_annualrates', con=engine, if_exists='append', index=False)

In [36]:
sub_coh_annual_temp_df.to_sql(name='subgroup_cohorts_annualrates', con=engine, if_exists='append', index=False)

#### The following tables were not a part of our original dataset and will be later populated in our database (the tables are already created via Step 2)

- students
- teachers
- student_advisors
- taught_by_history
- classrooms
- homerooms

<b> Since the tables listed above are under separate permissions due to privacy regulation, and they are unnecessary for funding selection, they will be appended post-analysis by an analytics team with CRM access for datapoints regarding individuals (such as teachers and students). <b>